x-pack/plugins/security_solution/public/common/components/markdown_editor/plugins/osquery/index.tsx

```diff
@@ -0,0 +1,236 @@
+ * 2.0.
+ */
+
+import { pickBy, isEmpty } from 'lodash';
+import type { Plugin } from 'unified';
+import React, { useContext, useMemo, useState, useCallback } from 'react';
+import type { RemarkTokenizer } from '@elastic/eui';
+import {
+  EuiSpacer,
+  EuiCodeBlock,
+  EuiModalHeader,
+  EuiModalHeaderTitle,
+  EuiModalBody,
+  EuiModalFooter,
+  EuiButton,
+  EuiButtonEmpty,
+} from '@elastic/eui';
+import { useForm, FormProvider } from 'react-hook-form';
+import styled from 'styled-components';
+import type { EuiMarkdownEditorUiPluginEditorProps } from '@elastic/eui/src/components/markdown_editor/markdown_types';
+import { useKibana } from '../../../../lib/kibana';
+import { LabelField } from './label_field';
+import OsqueryLogo from './osquery_icon/osquery.svg';
+import { OsqueryFlyout } from '../../../../../detections/components/osquery/osquery_flyout';
+import { BasicAlertDataContext } from '../../../event_details/investigation_guide_view';
+import { convertECSMappingToObject } from './utils';
+
+const StyledEuiButton = styled(EuiButton)`
+  > span > img {
+    margin-block-end: 0;
+  }
+`;
+
+const OsqueryEditorComponent = ({
+  node,
+  onSave,
+  onCancel,
+}: EuiMarkdownEditorUiPluginEditorProps<{
+  configuration: {
+    label?: string;
+    query: string;
+    ecs_mapping: { [key: string]: {} };
+  };
+}>) => {
+  const { osquery } = useKibana().services;
+  const formMethods = useForm<{
+    label: string;
+    query: string;
+    ecs_mapping: Record<string, unknown>;
+  }>({
+    defaultValues: {
+      label: node?.configuration?.label,
+      query: node?.configuration?.query,
+      ecs_mapping: node?.configuration?.ecs_mapping,
+    },
+  });
+
+  const onSubmit = useCallback(
+    (data) => {
+      onSave(
+        `!{osquery${JSON.stringify(
+          pickBy(
+            {
+              query: data.query,
+              label: data.label,
+              ecs_mapping: convertECSMappingToObject(data.ecs_mapping),
+            },
+            (value) => !isEmpty(value)
+          )
+        )}}`,
+        {
+          block: true,
+        }
+      );
+    },
+    [onSave]
+  );
+
+  const OsqueryActionForm = useMemo(() => {
+    if (osquery?.LiveQueryField) {
+      const { LiveQueryField } = osquery;
+
+      return (
+        <FormProvider {...formMethods}>
+          <LabelField />
+          <EuiSpacer size="m" />
+          <LiveQueryField formMethods={formMethods} />
+        </FormProvider>
+      );
+    }
+    return null;
+  }, [formMethods, osquery]);
+
+  return (
+    <>
+      <EuiModalHeader>
+        <EuiModalHeaderTitle>{`${node ? 'Edit' : 'Add'} Osquery query`}</EuiModalHeaderTitle>
+      </EuiModalHeader>
+
+      <EuiModalBody>
+        <>{OsqueryActionForm}</>
+      </EuiModalBody>
+
+      <EuiModalFooter>
+        <EuiButtonEmpty onClick={onCancel}>{'Cancel'}</EuiButtonEmpty>
+        <EuiButton onClick={formMethods.handleSubmit(onSubmit)} fill>
+          {'Attach query'}
```

Comment:
Confirm button label

Reply:
Can the button label be different for the Add and Edit modals.

For the Add modal, I suggest "Add query".

For the Edit modal, "Save query" or "Save changes"

Also here is an edit for the placeholder text:  Search for a query or write a new one.  We try to avoid using words like "below" for accessibility.

Reply:
I like @gchaps suggestions for modal/button. I would keep "label" because it could be different text from a saved query name.

Reply:
should internationalize these labels as well I think